File: todo-go-http/routes/todos.go
package routes

import (
	"dv4all/todoapi/pgdb"
	"dv4all/todoapi/response"
	"encoding/json"
	"log"
	"net/http"
	"strconv"
	"strings"
)

func handleTodos(res http.ResponseWriter, req *http.Request) {
	var data response.Response
	var listID int32
	upperMethod := strings.ToUpper(req.Method)
	// extract listID from params
	params := strings.Split(req.URL.Path, "/")
	// log.Println("Path: ", req.URL.Path)
	// log.Println("params len: ", len(params))
	// url = /todos/{listID}/items
	if len(params) == 4 {
		log.Println("Url listID: ", params[2])
		i64, err := strconv.ParseInt(params[2], 10, 32)
		if err != nil {
			data = response.SetErrorResponse(err, response.ServerStatus{
				Status:     400,
				StatusText: "Bad request",
			})
			data.ReturnResponse(res)
			return
		}
		listID = int32(i64)
	}
	switch upperMethod {
	case "GET":
		if listID > 0 {
			getAllTodoItems(req, res, listID)
		} else {
			getAllTodoLists(req, res)
		}
	case "POST":
		if listID > 0 {
			addTodoItem(req, res, listID)
		} else {
			addTodoList(req, res)
		}
	default:
		log.Printf("METHOD NOT SUPPORTED...%v", upperMethod)
		data.Status = http.StatusBadRequest
		data.StatusText = "Method not supported"
		data.Payload = "Method not supported"
		//return bad request
		data.ReturnResponse(res)
	}
}

func getAllTodoLists(req *http.Request, res http.ResponseWriter) {
	var data response.Response
	tls, err := pgdb.GetAllTodoLists()

	if err != nil {
		data = response.SetErrorResponse(err, response.ServerStatus{})
	} else {
		data = response.SetOKResponse(tls)
	}
	//write reponse
	data.ReturnResponse(res)
}

func getTodoListFromReqBody(req *http.Request, res http.ResponseWriter) (pgdb.BaseList, error) {
	var data response.Response
	var tl pgdb.BaseList
	err := json.NewDecoder(req.Body).Decode(&tl)
	if err != nil {
		log.Println("getTodoListFromReqBody: ", err)
		data = response.SetErrorResponse(err, response.ServerStatus{
			Status:     http.StatusBadRequest,
			StatusText: "Failed to extract data from request.Body",
		})
		data.ReturnResponse(res)
		//exit
		return tl, err
	}
	return tl, nil
}

func addTodoList(req *http.Request, res http.ResponseWriter) {
	var data response.Response
	input, err := getTodoListFromReqBody(req, res)
	if err != nil {
		//exit on error
		return
	}
	todolist, err := pgdb.AddTodoList(input)
	// check if insert failed
	if err != nil {
		data = response.SetErrorResponse(err, response.ServerStatus{
			Status:     http.StatusBadRequest,
			StatusText: "Failed to insert data into db",
		})
	} else {
		data = response.SetOKResponse(todolist)
	}
	//write reponse
	data.ReturnResponse(res)
}

func getAllTodoItems(req *http.Request, res http.ResponseWriter, lid int32) {
	items, err := pgdb.GetTodoItems(lid)
	if err != nil {
		data := response.SetErrorResponse(err, response.ServerStatus{
			Status:     http.StatusBadRequest,
			StatusText: "Failed to extract data from request.Body",
		})
		data.ReturnResponse(res)
	}
	data := response.SetOKResponse(items)
	data.ReturnResponse(res)
}

func getTodoItemFromBody(req *http.Request, res http.ResponseWriter) (pgdb.BaseTodoItem, error) {

	var td pgdb.BaseTodoItem
	err := json.NewDecoder(req.Body).Decode(&td)
	if err != nil {
		data := response.SetErrorResponse(err, response.ServerStatus{
			Status:     400,
			StatusText: "Failed to extract todo item from body",
		})
		data.ReturnResponse(res)
		return td, err
	}
	// data := response.SetOKResponse(td)
	// data.ReturnResponse(res)
	return td, nil
}

func addTodoItem(req *http.Request, res http.ResponseWriter, lid int32) {
	data, err := getTodoItemFromBody(req, res)
	if err != nil {
		return
	}
	todo, err := pgdb.AddTodoItem(lid, data)
	if err != nil {
		data := response.SetErrorResponse(err, response.ServerStatus{})
		data.ReturnResponse(res)
		return
	}
	resp := response.SetOKResponse(todo)
	resp.ReturnResponse(res)
}

File: todo-express-api/src/routes/todos.js
const db = require('../db/todos')
const resp = require('../utils/resp')

function getAllTodoLists(req, res){
  return db.GetAllTodoLists()
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function addTodoList(req, res){
  const todolist = req.body
  return db.AddTodoList(todolist)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function updateTodoList(req, res) {
  const {lid} = req.params
  const todolist = req.body
  return db.UpdateTodoList(lid,todolist)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function deleteTodoList(req, res){
  const {lid} = req.params
  return db.DeleteTodoList(lid)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function getTodoItems(req, res){
  const {lid} = req.params
  return db.GetTodoItems(lid)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function getTodoItem(req, res){
  const {id} = req.params
  return db.GetTodoItem(id)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function getTodoList(req, res){
  const {lid} = req.params
  return db.GetTodoList(lid)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data[0])
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}


function addTodoItem(req, res){
  // const {lid} = req.params
  const todo = req.body
  return db.AddTodoItem(todo)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function updateTodoItem(req, res) {
  const {id} = req.params
  const todo = req.body
  return db.UpdateTodoItem(id,todo)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

function deleteTodoItem(req, res){
  const {id} = req.params
  return db.DeleteTodoItem(id)
    .then(data=>{
      // console.log("data received...", data)
      resp.respOK(res,data)
    })
    .catch(err=>{
      resp.respErr(res,500,err.message)
    })
}

module.exports={
  getAllTodoLists,
  getTodoList,
  addTodoList,
  updateTodoList,
  deleteTodoList,
  getTodoItems,
  addTodoItem,
  getTodoItem,
  updateTodoItem,
  deleteTodoItem
}